R/sf.R

```diff
@@ -135,11 +135,38 @@ stat_sf <- function(mapping = NULL, data = NULL, geom = "rect",
       na.rm = na.rm,
       legend = if (is.character(show.legend)) show.legend else "polygon",
       ...
-    )
+    ),
+    layer_class = LayerSf
   )
 }
 
 
+# A special sf layer that auto-maps geometry data to the `geometry` aesthetic
+
+#' @export
+#' @rdname ggsf
+#' @usage NULL
+#' @format NULL
+LayerSf <- ggproto("LayerSf", Layer,
+  setup_layer = function(self, plot) {
+    # Automatically determine the name of the geometry column
+    # and add the mapping if it doesn't exist
+    data <- self$layer_data(plot$data)
+
+    if (!is.null(data) && is_sf(data)) {
```

Comment:
When the data is not specified either in the plot or the layer, `Layer$layer_data()` probably returns `waiver` instead of `NULL`. `ggplot()$data` is `waiver` thanks to `fortify.NULL()`. So, maybe `is.waive()` here?

But, I'm not sure if it makes sense to set `geometry` aes when we already know `data` doesn't have `geometry` column...

Reply:
I'll change it to `is.waive()`. I'm not sure about the addition of the `geometry` aes when there is no geometry data, I copied that part from the existing code.

Reply:
> I'm not sure about the addition of the geometry aes when there is no geometry data

I'm not sure, too... You know, with the current implementation, `geom_sf()` cannot know the geometry column when the `sf` object is specified in the plot, not in the layer. So, placing the last glimmer of hope on `"geometry"` may be a reasonable strategy.

But, with `setup_layer()`, maybe we can just `stop()` when the data is not `sf`?

Reply:
I don't even think we need to stop. We can just not do anything, and the general ggplot2 machinery will complain about a missing geometry aesthetic.

Reply:
Yes, plotting will fail in the either case. I'm OK with the current implementation, but, FWIW, my thoughts was:

`LayerSf` ensures the mapping has `geometry` aesthetic as long as the `data` is `sf`. In other words, if there's no `geometry`, `data` is probably not `sf`. In this case, specifying `geometry` won't fix the problem. So, maybe is the error message more helpful if it complains about the fact that `data` is not `sf`?